db/blip_sync_context.go

```diff
@@ -669,33 +671,44 @@ func toHistory(revisions Revisions, knownRevs map[string]bool, maxHistory int) [
 	return history
 }
 
-// reportIfTimeElapsed will report stats if enough time has passed since the previous report.
-func (s *blipSyncStats) reportIfTimeElapsed(bsc *BlipSyncContext) {
+// reportStatsIfTimeElapsed will report stats if enough time has passed since the previous report.
+func (bsc *BlipSyncContext) reportStatsIfTimeElapsed() {
 	currentTime := time.Now().UnixMilli()
-	if (currentTime - s.lastReportTime.Load()) > bsc.blipContextDb.BlipStatsReportingInterval {
-		s.lastReportTime.Store(currentTime)
-		s.report(bsc)
-
+	if !bsc.timeElapsedForStatsReporting(currentTime) {
+		return
 	}
+	bsc.reportStats(currentTime)
+
+}
+
+// timeElapsedForStatsReporting will return true if enough time has passed since the previous report.
+func (bsc *BlipSyncContext) timeElapsedForStatsReporting(currentTime int64) bool {
+	return (currentTime - bsc.stats.lastReportTime.Load()) > bsc.blipContextDb.Options.BlipStatsReportingInterval
 }
 
-// report will update the stats on a database immediately
-func (s *blipSyncStats) report(bsc *BlipSyncContext) {
+// reportStats will update the stats on a database immediately
+func (bsc *BlipSyncContext) reportStats(currentTime int64) {
 	if bsc.blipContextDb == nil || bsc.blipContext == nil {
 		return
 	}
 	dbStats := bsc.blipContextDb.DbStats.Database()
 	if dbStats == nil {
 		return
 	}
-	s.lock.Lock()
-	defer s.lock.Unlock()
+	bsc.stats.lock.Lock()
+	defer bsc.stats.lock.Unlock()
+
+	// a goroutine has already updated these stats
+	if !bsc.timeElapsedForStatsReporting(currentTime) {
```

Comment:
Can this check just use time.Now instead of currentTime?  Consider a case where many goroutines spawn with currentTime=t and queue up to acquire the lock.  At t+BlipStatsReportingInterval, all of those goroutines are going to try to update stats, even though the stats were just updated by the previous goroutine.

It feels like reportStats actually just needs to pass in a flag like 'forceUpdate bool' that bypasses the time check (that would be used on close).  With that approach I think you could get rid of reportIfTimeElapsed and just incorporate that logic into reportStats.